### PR TITLE
[D1] Run sema3 for all modules when generating deps

### PR DESCRIPTION
**Pull request by stefan-koch-sociomantic**
This makes sure we don't miss depdenecies introduces imports functions or unittests

### CODE REVIEW

**Comment by stefan-koch-sociomantic** [comment]
backport of: https://github.com/dlang/dmd/pull/6748

File: src/mars.c [hunk 1]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
     {
         assert(global.params.moduleDepsFile != NULL);
 
+        // when generating depdencies we have to run sema3
+        // because otherwise we will miss some depdencies
+        // introdcued in function bodies
+        for (size_t i = 0; i < Module::amodules.dim; i++)
```
Comment (RazvanN7) [inline src/mars.c]: Couldn't this start from 1 since the first module is the module being compiled and semantic has already been performed on it ?

**Review by RazvanN7 — APPROVED** [review]
Looks good.

File: src/mars.c (hunk 1 above)
Comment (stefan-koch-sociomantic) [reply src/mars.c]: It could but who cares. I'd rather give no guarantees about order.
That's how bugs are born :)

Comment (jacob-carlborg) [reply src/mars.c]: Please make it a `foreach` loop.

Comment (jacob-carlborg) [reply src/mars.c]: Never mind.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by WalterBright** [comment]
I object to this because:

* it is not fixing any known bug
* it will substantially slow down the compiler
* the compiler has moved towards lazily doing all sorts of analyses - for speed purposes
* if there are specific problems found where semantic3() isn't being run, that's where it should be run. There are many such places already in the compiler

File: src/mars.c [hunk 2]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
     {
         assert(global.params.moduleDepsFile != NULL);
 
+        // when generating depdencies we have to run sema3
```
Comment (WalterBright) [inline src/mars.c]: please don't abbreviate `semantic3` - such will make grep useless

also, `dependencies`

File: src/mars.c [hunk 3]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
...
+        // because otherwise we will miss some depdencies
```
Comment (WalterBright) [inline src/mars.c]: dependencies

File: src/mars.c [hunk 4]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
...
+        // introdcued in function bodies
```
Comment (WalterBright) [inline src/mars.c]: introduced

File: src/mars.c [hunk 5]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
...
+        for (size_t i = 0; i < Module::amodules.dim; i++)
+        {
+            m = Module::amodules[i];
+            if (global.params.verbose)
+                printf("[deps] semantic(2/3) %s\n", m->toChars());
+            // not sure if it can happen that we havn't run sema2 ?
+            // better safe then sorry, it does not cost more then a few cycles
```
Comment (WalterBright) [inline src/mars.c]: than

Need a better rationale than "not sure" and "better safe than sorry"

**Comment by mihails-strasuns-sociomantic** [comment]
@WalterBright this only targets dmd-1.x branch and is intended to fix problem already fixed in main D2 branch. Is that still a no-go?

**Comment by WalterBright** [comment]
Sorry, I missed that it was D1 only. It's good to go!

**Comment by ibuclaw** [comment]
Shouldn't the spelling fixes been done prior to merge? We should still try to keep a reasonable code quality regardless of the branch.

**Comment by mihails-strasuns-sociomantic** [comment]
Yes, I was intending to merge this myself after doing necessary manual testing (there are no auto-tests for D1 branch to rely on) ;)

Will make separate fixup PR if I find something.

**Comment by stefan-koch-sociomantic** [comment]
I am going to fix the spelling of the comments.
@WalterBright thanks for pointing them out

File: src/mars.c [hunk 6]
```diff
@@ -1531,6 +1531,20 @@ int main(int iargc, const char *argv[])
...
 
+        // when generating depdencies we have to run sema3
+        // because otherwise we will miss some depdencies
+        // introdcued in function bodies
+        for (size_t i = 0; i < Module::amodules.dim; i++)
+        {
+            m = Module::amodules[i];
+            if (global.params.verbose)
+                printf("[deps] semantic(2/3) %s\n", m->toChars());
+            // not sure if it can happen that we havn't run sema2 ?
```
Comment (gautam-kotian-sociomantic) [inline src/mars.c]: typo `havn't`

Comment (stefan-koch-sociomantic) [reply src/mars.c]: fixed in the fixup #8166

**Pull request merged**